### ISSUE TITLE
WM_CLASS and WM_NAME not set before window is mapped

### ISSUE THREAD

**Issue by jonhoo** [tomaka/glutin#879]
Many Linux window managers rely on being able to match against a new window's `WM_CLASS` or `WM_NAME` at launch to place them on the right workspace, set floating properties, etc. Since the window manager only looks at the value of these properties when the window is first mapped, ICCCM [requires](https://tronche.com/gui/x/icccm/sec-4.html#WM_CLASS) (for `WM_CLASS`) that
> This property must be present when the window leaves the Withdrawn state and may be changed only while the window is in the Withdrawn state. Window managers may examine the property only when they start up and when the window leaves the Withdrawn state, but there should be no need for a client to change its state dynamically.

While no such requirement exists for `WM_NAME`, setting it before the window is mapped is [advantageous](https://bbs.archlinux.org/viewtopic.php?pid=1581353#p1581353) for the same reason. Unfortunately, glutin currently does not appear to do this (neither does [Spotify](https://community.spotify.com/t5/Desktop-Linux-Windows-Web-Player/WM-CLASS-window-property-on-Linux/td-p/1247042) fwiw). It sets these properties *after* mapping the new window. While using `xprop` to inspect the window launched by `examples/window.rs` does show the right `WM_NAME` and `WM_CLASS` ("A fantastic window!"), this information is not visible to window managers. For example, neither of these two xmonad window hooks work:

```haskell
className =? "A fantastic window!" --> doShift "1"
```
```haskell
title =? "A fantastic window!" --> doShift "1"
```

On the other hand, matching against "" does move the window, suggesting that the window is initially mapped without title or class.

**Comment by jonhoo** [comment]
This may actually be a bug in winit, because it [calls `display.xlib.XMapRaised`](https://github.com/tomaka/winit/blob/master/src/platform/linux/x11/window.rs#L409) *before* it [sets name and class](https://github.com/tomaka/winit/blob/master/src/platform/linux/x11/window.rs#L469).

**Comment by jonhoo** [comment]
This is fixed in winit in tomaka/winit#168. I've manually verified that running glutin with winit including that patch correctly sets `WM_NAME` and `WM_CLASS` before mapping the window (and those attributes are therefore visible to the window manager).

**Comment by jonhoo** [comment]
@tomaka I don't think this is actually resolved until glutin pulls in an updated version of winit?
